File: src/java/FloydWarshall.java
/*
 * Floyd-Warshall algorithm in Java
 * All pairs shortest path algorithm
 * Time Complexity:  O(n³)
 * Space Complexity: O(n²)
*/

public class FloydWarshall {

  public static void showMatrix(long[][] matriz, int nroVertices) {
    for (int i = 0; i < nroVertices; i++) {
      for (int j = 0; j < nroVertices; j++) {
        if (matriz[i][j] < 10) {
          System.out.print(" " + matriz[i][j] + " ");
        } else {
          System.out.print(matriz[i][j] + " ");
        }
      }
      System.out.println();
    }
    System.out.println();
  }

  public static void main(String[] args) {
    int nroVertices = 5;
    long[][] matriz = {
      {0, 2, 10, 5, 7},
      {2, 0, 3, 3, 1},
      {10, 3, 0, 1, 2},
      {5, 3, 1, 0, Long.MAX_VALUE},
      {7, 1, 2, 2, 0}
    };

    // Display the original matrix
    System.out.println("Original matrix:");
    showMatrix(matriz, nroVertices);

    floydWarshall(matriz, nroVertices);

    // Display the updated matrix
    System.out.println("Updated matrix:");
    showMatrix(matriz, nroVertices);

    // Show all shortest paths
    System.out.println();
    for (int i = 0; i < nroVertices; i++) {
      for (int x = 0; x < nroVertices; x++) {
        System.out.println("Shortest distance from " + i + " to " + x + " = " + matriz[x][i] + ".");
      }
    }
    System.out.println();
  }

  public static void floydWarshall(long[][] matriz, int n) {
    for (int x = 0; x < n; x++) { // Intermediary vertex
      for (int y = 0; y < n; y++) { // Origin vertex
        for (int z = 0; z < n; z++) { // Destination vertex
          if (matriz[y][z] > (matriz[y][x] + matriz[x][z])) {
            matriz[y][z] = matriz[y][x] + matriz[x][z];
          }
        }
      }
    }
  }
}

File: src/c/FloydWarshall.c
/*
*	Grafos - Algoritmo de Floyd-Warshall em C
*	Complexidade: Teta de vértices ao cubo = Teta(n^3)
*
*	Encontra o caminho de todos para todos os vértices
*
*	Grafo com 5 vértices e 6 arestas
*
*			   6
*	(0)-----------------(1)
*	 |					 |
* 10 |					 | 2
*	 |		   1		 |
*	(2)-----------------(3)
*	  \					/
*	 3 \			   / 8
*		\			  /
*		 -----(4)-----			 
*
*	 Matriz de Distância
*	   0   1   2   3   4
*	0  0   6  10   -   -
*	1  6   0   -   2   -
*	2 10   -   0   1   3
*	3  -   2   1   0   8
*	4  -   -   3   8   0
*
* 	Para valores infinitos será considerado o valor: 4294967295
*
*/

#include <stdio.h>

#define nroVertices 5 // Define uma constante 5 que é a quantidade de vértices do grafo

// Algoritmo de Floyd-Warshall recebe como parâmetro a matriz de distância e o número de vértices
void FloydWarshall(unsigned long int matriz[nroVertices][nroVertices], int n){
	for(int x = 0; x < n; x++){
		for(int y = 0; y < n; y++){
			for(int z = 0; z < n; z++){
				if( matriz[y][z] > (matriz[y][x] + matriz[x][z]) )
					matriz[y][z] = matriz[y][x] + matriz[x][z];
			}	
		}	
	}
}

int main(){

	unsigned long int Matriz[nroVertices][nroVertices] = {{         0,          6,         10, 4294967295, 4294967295 },
														 {          6,          0, 4294967295,          2, 4294967295 },
														 {         10, 4294967295,          0,          1,          3 },
														 { 4294967295,          2,          1,          0,          8 },
														 { 4294967295, 4294967295,          3,          8,          0 }};

	FloydWarshall(Matriz, nroVertices);

	// Mostra a matriz com os valores atualizados
	printf("Matriz:\n");
	for (int i = 0; i < nroVertices; ++i){
		for (int e = 0; e < nroVertices; ++e){
			if( Matriz[i][e] < 10 )
				printf(" %lu ", Matriz[i][e]);
			else
				printf("%lu ", Matriz[i][e]);
		}
		printf("\n");
	}
	printf("\n");

	//printf("Total caminho mais curto do vertice 0 ao 4: %lu\n", Matriz[4][0]); // [destino] [origem]
	//printf("Total caminho mais curto do vertice 2 ao 0: %lu\n", Matriz[0][2]);
	//printf("Total caminho mais curto do vertice 4 ao 3: %lu\n\n", Matriz[3][4]);

	// Mostra todos os caminhos 
	printf("\n");
	for (int i = 0; i < nroVertices; ++i){
		for (int x = 0; x < nroVertices; ++x){
			printf("Menor distancia saindo do %d para chegar ao %d = %lu.\n", i, x, Matriz[x][i]);
		}
	}
	printf("\n\n");

	return 0;
}



//  In english 

/*
* Graphs - Floyd-Warshall Algorithm in C
* Complexity: Theta of vertices cubed = Theta(n^3)
*
* Finds the shortest path from all vertices to all vertices
*
* Graph with 5 vertices and 6 edges
*
*              6
*   (0)-----------------(1)
*    |                   |
* 10 |                   | 2
*    |         1         |
*   (2)-----------------(3)
*    \                  /
*   3 \                / 8
*      \              /
*       -----(4)-----             
*
* Distance Matrix
*     0   1   2   3   4
* 0  0   6  10   -   -
* 1  6   0   -   2   -
* 2 10   -   0   1   3
* 3  -   2   1   0   8
* 4  -   -   3   8   0
*
* For infinite values, the value will be considered: 4294967295
*
*/


// #define nroVertices 5 // Defines a constant 5 which is the number of vertices in the graph

// // The Floyd-Warshall algorithm receives the distance matrix and the number of vertices as parameters
// void FloydWarshall(unsigned long int matriz[nroVertices][nroVertices], int n) {
//     for (int x = 0; x < n; x++) {
//         for (int y = 0; y < n; y++) {
//             for (int z = 0; z < n; z++) {
//                 // If the current path is greater than the path through vertex x, update the distance
//                 if (matriz[y][z] > (matriz[y][x] + matriz[x][z])) {
//                     matriz[y][z] = matriz[y][x] + matriz[x][z];
//                 }
//             }    
//         }    
//     }
// }

// int main() {
//     // Initialize the distance matrix with distances between vertices
//     unsigned long int Matriz[nroVertices][nroVertices] = {
//         {0, 6, 10, 4294967295, 4294967295},
//         {6, 0, 4294967295, 2, 4294967295},
//         {10, 4294967295, 0, 1, 3},
//         {4294967295, 2, 1, 0, 8},
//         {4294967295, 4294967295, 3, 8, 0}
//     };

//     // Execute the Floyd-Warshall algorithm
//     FloydWarshall(Matriz, nroVertices);

//     // Print the updated distance matrix
//     printf("Matrix:\n");
//     for (int i = 0; i < nroVertices; ++i) {
//         for (int e = 0; e < nroVertices; ++e) {
//             if (Matriz[i][e] < 10)
//                 printf(" %lu ", Matriz[i][e]);
//             else
//                 printf("%lu ", Matriz[i][e]);
//         }
//         printf("\n");
//     }
//     printf("\n");

//     // Show all shortest paths
//     printf("\n");
//     for (int i = 0; i < nroVertices; ++i) {
//         for (int x = 0; x < nroVertices; ++x) {
//             printf("Shortest distance from %d to %d = %lu.\n", i, x, Matriz[x][i]);
//         }
//     }
//     printf("\n\n");

//     return 0;
// }
